File: apps/go/chapter01/basic_algorithms.go
// Package chapter01 第1章 基本的なアルゴリズム
package chapter01

import (
	"fmt"
	"strings"
)

// Max3 3つの整数値の最大値を返す
func Max3(a, b, c int) int {
	maximum := a
	if b > maximum {
		maximum = b
	}
	if c > maximum {
		maximum = c
	}
	return maximum
}

// Med3 3つの整数値の中央値を返す
func Med3(a, b, c int) int {
	if a >= b {
		if b >= c {
			return b
		} else if a <= c {
			return a
		} else {
			return c
		}
	} else if a > c {
		return a
	} else if b > c {
		return c
	} else {
		return b
	}
}

// JudgeSign 整数値の符号を判定する
func JudgeSign(n int) string {
	if n > 0 {
		return "その値は正です。"
	} else if n < 0 {
		return "その値は負です。"
	} else {
		return "その値は0です。"
	}
}

// Sum1ToNWhile while ループで 1 から n までの総和を求める
func Sum1ToNWhile(n int) int {
	total := 0
	i := 1
	for i <= n {
		total += i
		i++
	}
	return total
}

// Sum1ToNFor for ループで 1 から n までの総和を求める
func Sum1ToNFor(n int) int {
	total := 0
	for i := 1; i <= n; i++ {
		total += i
	}
	return total
}

// Alternative1 記号文字 '+' と '-' を交互に表示する（剰余判定方式）
func Alternative1(n int) string {
	var result strings.Builder
	for i := 0; i < n; i++ {
		if i%2 == 0 {
			result.WriteByte('+')
		} else {
			result.WriteByte('-')
		}
	}
	return result.String()
}

// Alternative2 記号文字 '+' と '-' を交互に表示する（パターン繰り返し方式）
func Alternative2(n int) string {
	result := strings.Repeat("+-", n/2)
	if n%2 != 0 {
		result += "+"
	}
	return result
}

// Rectangle 縦横が整数で面積が area の長方形の辺の長さを列挙する
func Rectangle(area int) string {
	var result strings.Builder
	for i := 1; i*i <= area; i++ {
		if area%i == 0 {
			result.WriteString(fmt.Sprintf("%dx%d ", i, area/i))
		}
	}
	return result.String()
}

// MultiplicationTable 九九の表を返す
func MultiplicationTable() string {
	var result strings.Builder
	result.WriteString(strings.Repeat("-", 27) + "\n")
	for i := 1; i <= 9; i++ {
		for j := 1; j <= 9; j++ {
			result.WriteString(fmt.Sprintf("%3d", i*j))
		}
		result.WriteByte('\n')
	}
	result.WriteString(strings.Repeat("-", 27))
	return result.String()
}

// TriangleLb 左下側が直角の二等辺三角形を返す
func TriangleLb(n int) string {
	var result strings.Builder
	for i := 1; i <= n; i++ {
		result.WriteString(strings.Repeat("*", i) + "\n")
	}
	return result.String()
}

File: apps/fsharp/Chapter01/BasicAlgorithms.fs
module BasicAlgorithms

open System.Text

/// 3つの整数値の最大値を返す
let max3 a b c =
    let mutable maximum = a
    if b > maximum then maximum <- b
    if c > maximum then maximum <- c
    maximum

/// 3つの整数値の中央値を返す
let med3 a b c =
    if a >= b then
        if b >= c then b
        elif a <= c then a
        else c
    elif a > c then a
    elif b > c then c
    else b

/// 整数値の符号を判定する
let judgeSign n =
    if n > 0 then "その値は正です。"
    elif n < 0 then "その値は負です。"
    else "その値は0です。"

/// while 相当で 1 から n までの総和を求める
let sum1ToNWhile n =
    let mutable total = 0
    let mutable i = 1
    while i <= n do
        total <- total + i
        i <- i + 1
    total

/// for 相当で 1 から n までの総和を求める
let sum1ToNFor n =
    let mutable total = 0
    for i in 1..n do
        total <- total + i
    total

/// 記号文字 '+' と '-' を交互に表示する（剰余判定方式）
let alternative1 n =
    let sb = StringBuilder()
    for i in 0..n-1 do
        sb.Append(if i % 2 <> 0 then '-' else '+') |> ignore
    sb.ToString()

/// 記号文字 '+' と '-' を交互に表示する（パターン繰り返し方式）
let alternative2 n =
    let sb = StringBuilder()
    for _ in 1..n/2 do
        sb.Append("+-") |> ignore
    if n % 2 <> 0 then sb.Append('+') |> ignore
    sb.ToString()

/// 縦横が整数で面積が area の長方形の辺の長さを列挙する
let rectangle area =
    let sb = StringBuilder()
    let mutable i = 1
    while i * i <= area do
        if area % i = 0 then
            sb.Append(sprintf "%dx%d " i (area / i)) |> ignore
        i <- i + 1
    sb.ToString()

/// 九九の表を返す
let multiplicationTable () =
    let sb = StringBuilder()
    sb.AppendLine(String.replicate 27 "-") |> ignore
    for i in 1..9 do
        for j in 1..9 do
            sb.Append(sprintf "%3d" (i * j)) |> ignore
        sb.AppendLine() |> ignore
    sb.Append(String.replicate 27 "-") |> ignore
    sb.ToString()

/// 左下側が直角の二等辺三角形を返す
let triangleLb n =
    let sb = StringBuilder()
    for i in 0..n-1 do
        sb.AppendLine(String.replicate (i + 1) "*") |> ignore
    sb.ToString()
